documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,366 @@
+  - PositiveInfinity
+  - NegativeInfinity
+  - NaN
+  - Epsilon
+  - NegativeZero
+  - Pi
+  - Tau
+  - E
+- Define the following static functions for these types to match IBinaryFloatingPointIeee754 (Vector replaced with type, e.g. `Vector2F<T>`) which returns a new vector with these operations applied component-wise, unless otherwise specified:
+  - Sqrt(Vector x)
+  - Acosh(Vector x)
+  - Asinh(Vector x)
+  - Atanh(Vector x)
+  - Cosh(Vector x)
+  - Sinh(Vector x)
+  - Tanh(Vector x)
+  - Acos(Vector x)
+  - AcosPi(Vector x)
+  - Asin(Vector x)
+  - AsinPi(Vector x)
+  - Atan(Vector x)
+  - AtanPi(Vector x)
+  - Cos(Vector x)
+  - CosPi(Vector x)
+  - Sin(Vector x)
+  - SinPi(Vector x)
+  - Tan(Vector x)
+  - TanPi(Vector x)
+  - DegreesToRadians(Vector degrees)
+  - RadiansToDegrees(Vector radians)
+  - SinCos(Vector x)
+    - Returns a tuple of 2 Vectors (Sin, Cos)
+  - SinCosPi(Vector x)
+    - Returns a tuple of 2 Vectors (SinPi, CosPi)
+  - Log(Vector x)
+  - Log(Vector x, Vector newBase)
+  - Log(Vector x, TScalar newBase)
+  - LogP1(Vector x)
+  - Log2(Vector x)
+  - Log2P1(Vector x)
+  - Log10(Vector x)
+  - Log10P1(Vector x)
+  - Exp(Vector x)
+  - ExpM1(Vector x)
+  - Exp2(Vector x)
+  - Exp2M1(Vector x)
+  - Exp10(Vector x)
+  - Exp10M1(Vector x)
+  - Pow(Vector x, Vector y)
+  - Pow(Vector x, TScalar y)
+  - Cbrt(Vector x)
+  - Hypot(Vector x, Vector y)
+  - Hypot(Vector x, TScalar y)
+  - RootN(Vector x, int n)
+  - Round(Vector x)
+  - Round(Vector x, int digits)
+  - Round(Vector x, MidpointRounding mode)
+  - Round(Vector x, int digits, MidpointRounding mode)
+  - Truncate(Vector x)
+  - Atan2(Vector x, Vector y)
+  - Atan2Pi(Vector x, Vector y)
+  - Atan2(Vector x, TScalar y)
+  - Atan2Pi(Vector x, TScalar y)
+  - BitDecrement(Vector x)
+  - BitIncrement(Vector x)
+  - FusedMultiplyAdd(Vector left, Vector right, Vector addend)
+  - FusedMultiplyAdd(Vector left, Vector right, TScalar addend)
+  - FusedMultiplyAdd(Vector left, TScalar right, Vector addend)
+  - FusedMultiplyAdd(Vector left, TScalar right, TScalar addend)
+  - ReciprocalEstimate(Vector x)
+  - ReciprocalSqrtEstimate(TVector x)
+  - ILogB(Vector x)
+    - Returns VectorNI<T>, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - ScaleB(Vector x, VectorNI<int> n)
+  - ScaleB(Vector x, int n)
+  - RoundToInt(Vector x)
+    - Returns `VectorNI`, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - FloorToInt(Vector x)
+    - Returns `VectorNI`, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
+  - CeilingToInt(Vector x)
+    - Returns `VectorNI`, where N matches the dimensionality of the vector
+    - **INFORMATIVE** This may require multiple methods depending on implementation
```

Comment:
suggest specifying that these should take a type parameter for the particular int scalar type to use

Reply:
There is a reason to have one that does and one that does not, which matches Maxine's implementation currently (int can be done implicitly, the others need casts I believe). This is an mainly an implementation detail, that I want to leave up to whoever is implementing it.